### PR TITLE
[Music]Fix audio player to start playback at offset or percentage

### PR DESCRIPTION
**Pull request by DaveTBlake**
Fix audio player to correctly start playback at given offset or percentage of track duration.

Previously audio would start playback from the given offset but incorrect times would be displayed by the OSD, percentage didn't work at all. This is now fixes and has been tested with both single track and cue sheet (multiple songs in one music file).

### CODE REVIEW

File: xbmc/cores/paplayer/PAPlayer.cpp [hunk 1]
```diff
@@ -312,7 +312,23 @@ bool PAPlayer::QueueNextFileEx(const CFileItem &file, bool fadeIn)
...
+
+  // Music from cuesheet => "item_start" and offset match
+  // Start offset defines where this song starts in file of multiple songs
+  if (si->m_fileItem.HasProperty("item_start") &&
+      (si->m_fileItem.GetProperty("item_start").asInteger() == si->m_fileItem.m_lStartOffset))
+  {
+    // Start stream at offset from cuesheet
+    si->m_startOffset = si->m_fileItem.m_lStartOffset;
+    starttime = 0; // No resume point
+  }
```
Comment (AlwinEsch) [inline xbmc/cores/paplayer/PAPlayer.cpp:329]: Maybe it makes sense to make the `double starttime = CUtil::ConvertMilliSecsToSecs(si->m_fileItem.m_lStartOffset);` into an `else`, as by `if(true)` it becomes changed to "0"?

File: xbmc/cores/paplayer/PAPlayer.cpp [hunk 2]
```diff
@@ -352,18 +368,15 @@ bool PAPlayer::QueueNextFileEx(const CFileItem &file, bool fadeIn)
...
   si->m_bytesPerFrame = si->m_bytesPerSample * si->m_audioFormat.m_channelLayout.Count();
   si->m_started = false;
   si->m_finishing = false;
   si->m_framesSent = 0;
   si->m_seekNextAtFrame = 0;
-  if (si->m_fileItem.HasProperty("audiobook_bookmark"))
-    si->m_seekFrame = si->m_audioFormat.m_sampleRate * CUtil::ConvertMilliSecsToSecs(si->m_fileItem.GetProperty("audiobook_bookmark").asInteger());
-  else
-    si->m_seekFrame = -1;
+  si->m_seekFrame = -1;
```
Comment (AlwinEsch) [inline xbmc/cores/paplayer/PAPlayer.cpp:379]: Maybe also to move this as else to new part below, as not needed double work then?

File: xbmc/cores/paplayer/PAPlayer.cpp [hunk 3]
```diff
@@ -374,6 +387,23 @@ bool PAPlayer::QueueNextFileEx(const CFileItem &file, bool fadeIn)
...
     streamTotalTime = si->m_endOffset - si->m_startOffset;
 
+  // Seek to a resume point
+  if (si->m_fileItem.HasProperty("StartPercent") &&
+      (si->m_fileItem.GetProperty("StartPercent").asDouble() > 0) &&
+      (si->m_fileItem.GetProperty("StartPercent").asDouble() <= 100))
+  {
+    si->m_seekFrame =
+        si->m_audioFormat.m_sampleRate *
+        CUtil::ConvertMilliSecsToSecs(static_cast<int>(+(static_cast<double>(
```
Comment (AlwinEsch) [inline xbmc/cores/paplayer/PAPlayer.cpp:397]: As a question to learn something new, what's the background of the `"+"` there. Is new to me and don't quite understand what it does :sweat_smile:?

**Review by AlwinEsch — COMMENTED** [review]
From overview it looks good to me, only the part with the else.

**Comment by AlwinEsch** [comment]
Ups already in, has clicked on the backward request to him and not looked about.

Thanks for the fix :+1:, always good to have and maybe it works then also better on audiodecoder add-ons where sometimes the seek not worked correct on tests.

File: xbmc/cores/paplayer/PAPlayer.cpp (hunk 3 above)
Comment (DaveTBlake) [reply xbmc/cores/paplayer/PAPlayer.cpp:397]: An interesting typo, no idea how that got in there and then I am blind to it. At least it does no harm.

File: xbmc/cores/paplayer/PAPlayer.cpp (hunk 1 above)
Comment (DaveTBlake) [reply xbmc/cores/paplayer/PAPlayer.cpp:329]: I implelented the logic in this way around to highlight that the cuesheet use of offset is the minority use.

**Comment by DaveTBlake** [comment]
Yes @AlwinEsch both addons and JSON calls should now be able to resume audio in right place and have player report duration etc. correctly. Thanks for your interest.

**Pull request merged**